docs/designs/keras_callback.md

```diff
@@ -0,0 +1,301 @@
+        super(LearningRateScheduler, self).__init__()
+        self.schedule_fn = schedule_fn
+
+    def on_train_batch_begin(self, batch, logs=None):
+        """
+        Args:
+            batch: integer, the model version requested from PS.
+            logs: dict. Has keys batch and size representing the current batch number and the size of the batch.
+        """
+        if not hasattr(self.model.optimizer, 'lr'):
+            raise ValueError('Optimizer must have a "lr" attribute.')
+        lr = self.schedule_fn(batch)
+        K.set_value(self.model.optimizer.lr, lr)
+```
+Using ParameterServerStrategy, the worker calculates the batch gradients and set gradients to PS. PS updates weights using optimizer after receiving gradients. Although the worker can call `on_train_batch_begin` in `LearningRateScheduler` to adjust the learning rate of optimizer in its model, we should send the learning rate with gradients to PS by GRPC and PS updates weights using the received learning rate.
+
+* Case 3: Callback to write additional summaries to TensorBoard after each evaluation job completes.
+```python
+class SummaryWriter(tf.keras.callbacks.Callback):
+    def on_test_end(self, logs=None):
+        """Call on the test job end
+        Args:
+            logs: dict. Has key metrics representing the evaluation metrics on the validation data.
+        """
+        metrics = logs.get("metrics", None)
+        if metrics is None:
+            return
+        write(metrics)
+```
+The master determine whether or not the evaluation job is completed by `EvaluationService.complete_task()`. So, we can call `on_test_end` After `EvaluationService.complete_task()` returns evaluation metrics.
+```python
+if evaluation_task_completed:
+    eval_metrics = self._evaluation_service.complete_task()
+    if eval_metrics is not None:
+        logs = {"metrics": eval_metrics}
+        self._callbacks_list.on_test_end(logs)
+```
+
+* Case 4: Callback to perform early stopping when the metrics are met after an evaluation job.
+```python
+def stop_fn(metrics_list):
+    """Function to determine whether or not to stop training
+    Args:
+        metrics_list: List to contain metrics of each evaluation job.
+    Retrun:
+        boolean: Stop training if true.
+    """
+    latest_metrics = metrics_list[-1]
+    return True if latest_metrics['auc'] > 0.8 else False
+
+class EarlyStopper(tf.keras.callbacks.Callback):
+    def __init__(self, stop_fn):
+        self.stop_train = stop_train
+        self.metrics_list = []
+        super(EarlyStopper, self).__init__()
+
+    def on_test_end(self, logs=None):
+        """Call on the test job end
+        Args:
+            logs: dict. Has key metrics representing the evaluation metrics on the validation data.
+        """
+        metrics = logs.get("metrics", None)
+        if metrics is None:
+            return
+        self.metrics_list.append(metrics)
+        self.model.stop = stop_train(self.metrics_list)
```

Comment:
does `model` contains a property named `stop`?

Reply:
The default properties of `tf.keras.Model` does not contain `stop`.
`tf.keras.callbacks.EarlyStopping` will set the  property`stop_training` 
https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/callbacks.py#L1268-L1270

The `stop_training` property of `model` is initialized before model iteration. https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/keras/engine/training_arrays.py#L260

We also can set the property in ElasticDL and check whether or not to stop.